packages/fetchai/connections/p2p_libp2p/aea/dhtpeer/dhtpeer.go

```diff
@@ -0,0 +1,818 @@
+/* -*- coding: utf-8 -*-
+* ------------------------------------------------------------------------------
+*
+*   Copyright 2018-2019 Fetch.AI Limited
+*
+*   Licensed under the Apache License, Version 2.0 (the "License");
+*   you may not use this file except in compliance with the License.
+*   You may obtain a copy of the License at
+*
+*       http://www.apache.org/licenses/LICENSE-2.0
+*
+*   Unless required by applicable law or agreed to in writing, software
+*   distributed under the License is distributed on an "AS IS" BASIS,
+*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*   See the License for the specific language governing permissions and
+*   limitations under the License.
+*
+* ------------------------------------------------------------------------------
+ */
+
+package dhtpeer
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/multiformats/go-multihash"
+
+	circuit "github.com/libp2p/go-libp2p-circuit"
+	kaddht "github.com/libp2p/go-libp2p-kad-dht"
+	routedhost "github.com/libp2p/go-libp2p/p2p/host/routed"
+
+	proto "github.com/golang/protobuf/proto"
+
+	aea "libp2p_node/aea"
+)
+
+// panics if err is not nil
+func check(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+func ignore(err error) {
+	if err != nil {
+		log.Println("TRACE", err)
+	}
+}
+
+// DHTPeer A full libp2p node for the Agents Communication Network.
+// It is required to have a local address and a public one
+// and can acts as a relay for `DHTClient`.
+// Optionally, it provides delegate service for tcp clients.
+type DHTPeer struct {
+	host         string
+	port         uint16
+	publicHost   string
+	publicPort   uint16
+	delegatePort uint16
+	enableRelay  bool
+
+	key             crypto.PrivKey
+	publicKey       crypto.PubKey
+	localMultiaddr  multiaddr.Multiaddr
+	publicMultiaddr multiaddr.Multiaddr
+	bootstrapPeers  []peer.AddrInfo
+
+	dht         *kaddht.IpfsDHT
+	routedHost  *routedhost.RoutedHost
+	tcpListener net.Listener
+
+	addressAnnounced bool
+	myAgentAddress   string
+	myAgentReady     func() bool
+	dhtAddresses     map[string]string
+	tcpAddresses     map[string]net.Conn
+	processEnvelope  func(aea.Envelope) error
+}
+
+// New creates a new DHTPeer
+func New(opts ...Option) (*DHTPeer, error) {
```

Comment:
Main comment: throughout you have a lot of "magic values". It would be better to declare them as constants on the top or make them configurable (if applicable).

Reply:
Agree, although some of them are more subject to change then others